legate/core/runtime.py

```diff
@@ -69,6 +69,87 @@
 T = TypeVar("T")
 
 
+class MachineModel:
+    def __init__(
+        self,
+        num_nodes: int,
+        num_gpus: int,
+        num_cpus: int,
+        num_omps: int,
+        legion_runtime: Optional[legion.legion_runtime_t] = None,
+        legion_context: Optional[legion.legion_context_t] = None,
+    ) -> None:
+        self._num_nodes = num_nodes
+        self._num_gpus = num_gpus
+        self._num_cpus = num_cpus
+        self._num_omps = num_omps
+        self._legion_runtime = legion_runtime
+        self._legion_context = legion_context
+
+        self._last_point = 0
+        if self._num_gpus > 0:
+            self._num_devices = self._num_gpus
+        elif self._num_omps > 0:
+            self._num_devices = self._num_omps
+        else:
+            self._num_devices = self._num_cpus
+
+    def __str__(self) -> str:
+        res = f"MachineModel({self._num_nodes}, {self._num_cpus}, "
+        res += f"{self._num_gpus}, {self._num_omps})"
+        return res
+
+    @property
+    def num_nodes(self) -> int:
+        return self._num_nodes
+
+    @property
+    def num_cpus(self) -> int:
+        return self._num_cpus
+
+    @property
+    def num_gpus(self) -> int:
+        return self._num_gpus
+
+    @property
+    def num_omps(self) -> int:
+        return self._num_omps
+
+    def next_available_point(self) -> DeviceID:
+        if self._last_point >= (self._num_devices - 1):
+            self._last_point = 0
+        else:
+            self._last_point += 1
+        return DeviceID(self._last_point)
+
+    def sharding_space(self) -> Union[IndexSpace, None]:
+        if (
+            self._legion_context is not None
+            and self._legion_runtime is not None
+        ):
+            rect = Rect([self._num_devices], [0], True, 1)
+            handle = legion.legion_index_space_create_domain(
+                self._legion_runtime, self._legion_context, rect.raw()
+            )
+            result = IndexSpace(
+                self._legion_context, self._legion_runtime, handle=handle
+            )
+            return result
+        else:
+            return None
+
+
+class DeviceID:
```

Comment:
@magnatelee I could have used `int` instead of DeviceID, if you think it would be better. I thought it would be nice to have a separate class for it so code is easier to understand/read.

Reply:
I feel a type alias would be sufficient if readability was the goal. and if we mean to add an abstraction for devices, I want it to contain more metadata than just the id to be useful. (for example, #439 might need such an abstraction.) for this PR, I don't think that abstraction would be necessary though.